marker_api/src/ast/common/span.rs

```diff
@@ -2,133 +2,469 @@ use std::marker::PhantomData;
+/// creation. If your lint is targeting code from macro expansions, please
+/// consider that users might not be able to influence the generated code. It's
+/// also worth checking that all linted nodes originate from the same macro expansion.
+/// Check out the documentation of [`ExpnInfo`].
 #[repr(C)]
-#[derive(Debug, Clone)]
+#[derive(Clone)]
 pub struct Span<'ast> {
-    source: &'ast SpanSource<'ast>,
-    /// The start marks the first byte in the [`SpanSource`] that is included in this
-    /// span. The span continues until the end position.
-    start: usize,
-    end: usize,
+    _lifetime: PhantomData<&'ast ()>,
+    /// The source of this [`Span`]. The id space and position distribution is
+    /// decided by the driver. To get the full source information it might be
+    /// necessary to also pass the start and end position to the driver.
+    source_id: SpanSrcId,
+    /// This information could also be retrieved, by requesting the [`ExpnInfo`]
+    /// of this span. However, from looking at Clippy and rustc lints, it looks
+    /// like the main interested is, if this comes from a macro expansion, not from
+    /// which one. Having this boolean flag will be sufficient to answer this simple
+    /// question and will save on extra [`SpanSrcId`] mappings.
+    from_expansion: bool,
+    start: SpanPos,
+    end: SpanPos,
+}
+
+impl<'ast> std::fmt::Debug for Span<'ast> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        fn fmt_pos(pos: Option<FilePos<'_>>) -> String {
+            match pos {
+                Some(pos) => format!("{}:{}", pos.line(), pos.column()),
+                None => "[invalid]".to_string(),
+            }
+        }
+
+        let src = self.source();
+        let name = match src {
+            SpanSource::File(file) => format!(
+                "{}:{} - {}",
+                file.file(),
+                fmt_pos(file.to_file_pos(self.start)),
+                fmt_pos(file.to_file_pos(self.end))
+            ),
+            SpanSource::Macro(expn) => format!("[Inside Macro] {:#?}", expn.call_site()),
+        };
+        f.debug_struct(&name).finish()
+    }
 }
 
 impl<'ast> Span<'ast> {
-    pub fn is_from_file(&self) -> bool {
-        matches!(self.source, SpanSource::File(..) | SpanSource::Sugar(..))
+    /// Returns `true`, if this [`Span`] comes from a macro expansion.
+    pub fn is_from_expansion(&self) -> bool {
+        self.from_expansion
+    }
+
+    /// Returns the code snippet that this [`Span`] refers to or [`None`] if the
+    /// snippet is unavailable.
+    ///
+    /// ```ignore
+    /// let variable = 15_000;
+    /// //             ^^^^^^
+    /// //             lit_span
+    ///
+    /// lit_span.snippet(); // -> Some("15_000")
+    /// ```
+    ///
+    /// There are several reasons, why a snippet might be unavailable. Also
+    /// depend on the used driver. You can also checkout the other snippet
+    /// methods to better deal with these cases:
```

Comment:
One example would be Rust-Analyzer, they wanted to have a stable AST representation, even before I started Marker. There is nothing really stopping them from also creating a driver. Complex lints should still use a compiler as a driver, but simple lints can in theory also be used in IDEs.

RA can also use Marker as a driver, like it can do for Clippy. But that would mean that the lint crates are reloaded every time. Personally, I also disabled that part of RA and always check in the console. 

Then it's just a general thing. If rustc was the only driver, it could have just wrapped pointers to rustc types. Looking back, that would have been better to have a minimal working prototype sooner, but this way there is a clear separation between the driver an API. Several internal types(mostly IDs) are based on rustc, but there is nothing stopping the development of another driver. 

cc: https://github.com/rust-marker/marker/issues/32